waiter/src/waiter/state.clj

```diff
@@ -31,44 +31,75 @@
+
 (defn- slots-available?
   "Predicate on healthy instances with available slots greater than used slots."
-  [{:keys [slots-assigned slots-used status-tags]}]
-  (and (healthy? status-tags) (> slots-assigned slots-used)))
+  [{:keys [slots-assigned slots-used] :as state}]
+  (and (healthy? state) (> slots-assigned slots-used)))
+
+(defn- only-lingering-requests?
+  "Returns truthy value if the instance is idle or processing expired requests based on the earliest-request-threshold-time"
+  [request-id->use-reason-map earliest-request-threshold-time]
+  (every? #(t/before? % earliest-request-threshold-time)
+          (->> request-id->use-reason-map
+               vals
+               (map :time))))
 
 (defn- killable?
-  "Only instances without slots used and not killed|locked can be killed."
-  [{:keys [slots-used status-tags]}]
-  (and (zero? slots-used)
-       (not-any? #(contains? status-tags %) [:killed :locked])))
+  "Only the following instances can be killed:
+   - without slots used and not killed|locked, or
+   - expired and idle or processing lingering requests"
+  [request-id->use-reason-map earliest-request-threshold-time {:keys [slots-used status-tags] :as state}]
+  (or (and (zero? slots-used)
+           (not-any? #(contains? status-tags %) [:killed :locked]))
+      (and (expired? state)
+           (only-lingering-requests? request-id->use-reason-map earliest-request-threshold-time))))
+
+(defn- find-max
+  "Uses the provided comparison function to find the max element in collection."
+  [comparison-fn collection]
+  (reduce #(if (pos? (comparison-fn %1 %2)) %1 %2) (first collection) (rest collection)))
 
 (defn find-killable-instance
   "While killing instances choose using the following logic:
    if there are expired and starting instances, only kill unhealthy instances that are not starting;
    choose instances in following order:
    - choose amongst idle instances
-   - choose youngest instance to break ties
+   - choose youngest healthy instance or oldest expired instance to break ties
    - choose idle unhealthy ones with highest priority
    - next, choose healthy, but blacklisted instances
-   - next, choose expired instance
+   - next, choose expired instances (idle or processing only lingering requests)
    - finally, choose healthy instances."
-  [id->instance instance-id->state acceptable-instance-id?]
-  (let [find-max (fn [comparison-fn collection]
-                   (reduce
-                     #(if (pos? (comparison-fn %1 %2)) %1 %2)
-                     (first collection)
-                     (rest collection)))
-        instance-id-state-pair->categorizer-vec (fn [[_ {:keys [status-tags]}]]
+  [id->instance instance-id->state acceptable-instance-id? instance-id->request-id->use-reason-map lingering-request-threshold-ms]
+  (let [earliest-request-threshold-time (t/minus (t/now) (t/millis lingering-request-threshold-ms))
+        extract-alphanumeric (fn [s] (str/replace (str s) #"[^A-Za-z0-9]" ""))
```

Comment:
The requirement on the `started-at` is that it is a string that must be formatted to allow string comparisons. So I'm relying on the ordering of alphanumeric characters.